File: locomotion/src/communication/src/QuadROSComm.cpp
#include "QuadROSComm.hpp"

QuadROSComm::QuadROSComm() : m_name("svan"), rclcpp::Node("svan"), CommunicationManager(DATA_ACCESS_MODE::EXECUTOR_TO_PLANT) {
    InitClass();
}

QuadROSComm::QuadROSComm(const std::string& name) : m_name(name), rclcpp::Node(name), CommunicationManager(name, DATA_ACCESS_MODE::EXECUTOR_TO_PLANT) {
    InitClass();
}

QuadROSComm::~QuadROSComm() {
    rclcpp::shutdown();
}

void QuadROSComm::InitClass() {
    m_dt = 1./m_loop_rate;

    m_joint_state_pub = this->create_publisher<sensor_msgs::msg::JointState>("/" + m_name + "/joint_cmd", 1);
    auto pub_timer = std::chrono::duration<double>(m_dt);
    m_timer = this->create_wall_timer(pub_timer, std::bind(&QuadROSComm::timer_cb, this));
    
    m_joint_state_sub = this->create_subscription<sensor_msgs::msg::JointState>(
        "/" + m_name + "/joint_state", 
        1, 
        std::bind(&QuadROSComm::get_joint_state_cb, this, std::placeholders::_1)
    );
    m_imu_sub = this->create_subscription<sensor_msgs::msg::Imu>(
        "/" + m_name + "/imu",
        1,
        std::bind(&QuadROSComm::get_imu_cb, this, std::placeholders::_1)
    );
    m_odom_sub = this->create_subscription<nav_msgs::msg::Odometry>(
        "/" + m_name + "/odom",
        1,
        std::bind(&QuadROSComm::get_odom_cb, this, std::placeholders::_1)
    );

    js_data = sensor_msgs::msg::JointState();
    js_data.position.clear();
    js_data.velocity.clear();
    js_data.effort.clear();
    for (int i = 0; i < 12; ++i) {
        js_data.position.push_back(0);
        js_data.velocity.push_back(0);
        js_data.effort.push_back(0);
    }
}

void QuadROSComm::get_joint_state_cb(const sensor_msgs::msg::JointState::SharedPtr msg) const {
    for (int i = 0; i < 12; ++i) {
        m_sensor_data_ptr -> q(i) = msg -> position[i];
        m_sensor_data_ptr -> qd(i) = msg -> velocity[i];
        m_sensor_data_ptr -> tau(i) = msg -> effort[i];
    }
}

void QuadROSComm::get_imu_cb(const sensor_msgs::msg::Imu::SharedPtr msg) const {
    m_sensor_data_ptr -> quat(0) = msg->orientation.x;
    m_sensor_data_ptr -> quat(1) = msg->orientation.y;
    m_sensor_data_ptr -> quat(2) = msg->orientation.z;
    m_sensor_data_ptr -> quat(3) = msg->orientation.w;

    m_sensor_data_ptr -> w_B(0) = msg->angular_velocity.x;
    m_sensor_data_ptr -> w_B(1) = msg->angular_velocity.y;
    m_sensor_data_ptr -> w_B(2) = msg->angular_velocity.z;
    
    m_sensor_data_ptr -> a_B(0) = msg->linear_acceleration.x;
    m_sensor_data_ptr -> a_B(1) = msg->linear_acceleration.y;
    m_sensor_data_ptr -> a_B(2) = msg->linear_acceleration.z;

    // TODO: populate a_W and w_W as well
}

void QuadROSComm::get_odom_cb(const nav_msgs::msg::Odometry::SharedPtr msg) const {
    // TODO: Use Odom data as required. Maybe extend sensor_data to have odom data?
}

void QuadROSComm::timer_cb() {
    for (int i = 0; i < 12; ++i) {
        js_data.position[i] = m_joint_command_data_ptr -> q(i);
        js_data.velocity[i] = m_joint_command_data_ptr -> qd(i);
        js_data.effort[i] = m_joint_command_data_ptr -> tau(i);
    }
    m_joint_state_pub->publish(js_data);
}

void QuadROSComm::Run() {
    // rclcpp::init(0, nullptr);
    rclcpp::spin(shared_from_this());
}

/***********************************************************/

CommROSP2E::CommROSP2E() : m_name("svan"), rclcpp::Node("svan"), CommunicationManager(DATA_ACCESS_MODE::PLANT_TO_EXECUTOR) {
    InitClass();
}

CommROSP2E::CommROSP2E(const std::string& name) : m_name(name), rclcpp::Node(name), CommunicationManager(name, DATA_ACCESS_MODE::PLANT_TO_EXECUTOR) {
    InitClass();
}

CommROSP2E::~CommROSP2E() {
    rclcpp::shutdown();
}

void CommROSP2E::InitClass() {
    m_dt = 1./m_loop_rate;

    m_joint_state_pub = this->create_publisher<sensor_msgs::msg::JointState>("/" + m_name + "/joint_state", 1);
    m_imu_pub = this->create_publisher<sensor_msgs::msg::Imu>("/" + m_name + "/imu", 1);
    m_odom_pub = this->create_publisher<nav_msgs::msg::Odometry>("/" + m_name + "/odom", 1);

    auto pub_timer = std::chrono::duration<double>(m_dt);
    m_timer = this->create_wall_timer(pub_timer, std::bind(&CommROSP2E::timer_cb, this));
    
    m_joint_cmd_sub = this->create_subscription<sensor_msgs::msg::JointState>(
        "/" + m_name + "/joint_cmd", 
        1, 
        std::bind(&CommROSP2E::get_joint_cmd_cb, this, std::placeholders::_1)
    );

    js_data = sensor_msgs::msg::JointState();
    imu_data = sensor_msgs::msg::Imu();
    odom_data = nav_msgs::msg::Odometry();

    js_data.position.clear();
    js_data.velocity.clear();
    js_data.effort.clear();
    for (int i = 0; i < 12; ++i) {
        js_data.position.push_back(0);
        js_data.velocity.push_back(0);
        js_data.effort.push_back(0);
    }
}

void CommROSP2E::get_joint_cmd_cb(const sensor_msgs::msg::JointState::SharedPtr msg) const {
    for (int i = 0; i < 12; ++i) {
        m_joint_command_data_ptr -> q(i) = msg -> position[i];
        m_joint_command_data_ptr -> qd(i) = msg -> velocity[i];
        m_joint_command_data_ptr -> tau(i) = msg -> effort[i];
    }
}

void CommROSP2E::timer_cb() {
    for (int i = 0; i < 12; ++i) {
        js_data.position[i] = m_sensor_data_ptr -> q(i);
        js_data.velocity[i] = m_sensor_data_ptr -> qd(i);
        js_data.effort[i] = m_sensor_data_ptr -> tau(i);
    }
    // Populate imu_data
    imu_data.orientation.w = m_sensor_data_ptr -> quat(3);
    imu_data.orientation.x = m_sensor_data_ptr -> quat(0);
    imu_data.orientation.y = m_sensor_data_ptr -> quat(1);
    imu_data.orientation.z = m_sensor_data_ptr -> quat(2);
    imu_data.linear_acceleration.x = m_sensor_data_ptr -> a_B(0);
    imu_data.linear_acceleration.y = m_sensor_data_ptr -> a_B(1);
    imu_data.linear_acceleration.z = m_sensor_data_ptr -> a_B(2);
    imu_data.angular_velocity.x = m_sensor_data_ptr -> w_B(0);
    imu_data.angular_velocity.y = m_sensor_data_ptr -> w_B(1);
    imu_data.angular_velocity.z = m_sensor_data_ptr -> w_B(2);
    // Populdate odometry data

    // Publish stuff
    m_joint_state_pub->publish(js_data);
    m_imu_pub->publish(imu_data);
    m_odom_pub->publish(odom_data);
}

void CommROSP2E::Run() {
    // rclcpp::init(0, nullptr);
    rclcpp::spin(shared_from_this());
}

File: locomotion/src/execution/src/QuadROSComm.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import JointState
from sensor_msgs.msg import Imu
from nav_msgs.msg import Odometry
import numpy as np
import threading

class JointData:
    def __init__(self):
        self.q = np.array([0.] * 12)
        self.qd = np.array([0.] * 12)
        self.tau = np.array([0.] * 12)
        self.kp = np.array([0.] * 12)
        self.kd = np.array([0.] * 12)

class SensorData(JointData):
    def __init__(self):
        self.quat = np.array([0., 0., 0., 1.])
        self.a_B = np.array([0., 0., 0.])
        self.w_B = np.array([0., 0., 0.])
        self.a_W = np.array([0., 0., 0.])
        self.w_W = np.array([0., 0., 0.])

class QuadROSComm(Node):
    def __init__(self, name):
        super().__init__(name)
        self.m_name = name
        self.m_dt = 0.001
        self.thread = None
        self.initClass()

    def setLoopRate(self, rate):
        self.m_loop_rate = rate
        self.m_dt = 1./rate
    
    def initClass(self):
        self.joint_cmd = JointState()
        # self.joint_cmd.position.clear()
        # self.joint_cmd.velocity.clear()
        # self.joint_cmd.effort.clear()
        self.joint_cmd.position = [0.] * 12
        self.joint_cmd.velocity = [0.] * 12
        self.joint_cmd.effort = [0.] * 12
        print("Joint cmd init: ", self.joint_cmd.position)

        self.joint_state = JointState()
        # self.joint_state.position.clear()
        # self.joint_state.velocity.clear()
        # self.joint_state.effort.clear()
        self.joint_state.position = [0.] * 12
        self.joint_state.velocity = [0.] * 12
        self.joint_state.effort = [0.] * 12

        self.imu = Imu()
        self.sensor_data = SensorData()
        self.joint_state_sub = self.create_subscription(
            JointState,
            "/" + self.m_name + "/joint_state",
            self.get_joint_state_cb,
            1)
        self.joint_state_sub
        self.imu_sub = self.create_subscription(
            Imu,
            "/" + self.m_name + "/imu",
            self.get_imu_cb,
            1)
        self.imu_sub
        self.joint_cmd_pub = self.create_publisher(
            JointState,
            "/" + self.m_name + "/joint_cmd",
            1)
        self.timer = self.create_timer(self.m_dt, self.timer_callback)
    
    def set_joint_command(self, jcmd):
        # print("Joint position current: ", self.joint_cmd.position)
        # print("Joint position command: ", jcmd.q)
        self.joint_cmd.position = list(jcmd.q)
        self.joint_cmd.velocity = list(jcmd.qd)
        self.joint_cmd.effort = list(jcmd.tau)
    
    def timer_callback(self):
        self.joint_cmd_pub.publish(self.joint_cmd)
    
    def get_joint_state_cb(self, msg):
        self.joint_state = msg

    def get_imu_cb(self, msg):
        self.imu = msg

    def getSensorData(self):
        self.sensor_data.q = np.array(self.joint_state.position)
        self.sensor_data.qd = np.array(self.joint_state.velocity)
        self.sensor_data.tau = np.array(self.joint_state.effort)
        self.sensor_data.quat[0] = self.imu.orientation.x
        self.sensor_data.quat[1] = self.imu.orientation.y
        self.sensor_data.quat[2] = self.imu.orientation.z
        self.sensor_data.quat[3] = self.imu.orientation.w
        # TODO: update the sensor data object as required. Currently only this much is needed to get observations for policy
        return self.sensor_data
    
    def getJointData(self):
        return {'q': self.joint_state.position, 'qd': self.joint_state.velocity}
    def getImuData(self):
        return {}
    
    def run(self):
        try:
            rclpy.spin(self)
        except KeyboardInterrupt:
            pass
        self.thread.join()

    def start_thread(self):
        self.thread = threading.Thread(target=self.run, daemon=True)
        self.thread.start()